File: triples/gui/linebreak.py
import re

# --- Helper functions for parsing LaTeX (mostly unchanged from previous) ---
DELIM_PAIRS = {
    "\\left(": "\\right)", "\\left[": "\\right]", "\\left{": "\\right}",
    "\\left|": "\\right|", "\\left.": "\\right.",
    "(": ")",
    "[": "]",
}
OPENER_REGEX_PATTERNS = [
    r"\\left\(", r"\\left\[", r"\\left\{", r"\\left\|", r"\\left\.",
    r"\\frac",
    r"\(", r"\[",
]
OPENING_BRACKETS_FOR_DEPTH = ['(', '[', '{']
CLOSING_BRACKETS_FOR_DEPTH = [')', ']', '}']
ADDITIVE_TERMS_LEFT_DELIMS = {
    "\\left(": "\\right)", "\\left[": "\\right]", "\\left{": "\\right}",
    "\\left|": "\\right|", "\\left.": "\\right."
}

def _get_latex_argument_content(text, start_pos):
    if start_pos >= len(text): return "", start_pos
    char = text[start_pos]
    if char == '{':
        brace_depth = 0
        for i in range(start_pos, len(text)):
            if text[i] == '{' and (i == 0 or text[i-1] != '\\'): brace_depth += 1
            elif text[i] == '}' and (i == 0 or text[i-1] != '\\'):
                brace_depth -= 1
                if brace_depth == 0: return text[start_pos + 1 : i], i + 1
        raise ValueError(f"Unmatched brace starting at {start_pos} in '{text}'")
    elif char == '\\':
        match = re.match(r"\\[a-zA-Z]+(?:\s*[*!]?)?", text[start_pos:]) or \
                re.match(r"\\.", text[start_pos:])
        if match:
            token = match.group(0)
            return token, start_pos + len(token)
        return text[start_pos], start_pos + 1
    return text[start_pos], start_pos + 1

# --- Part 1: Identify Additive Expressions (unchanged from previous correct version) ---
def get_additive_terms(latex_str):
    terms = []
    buffer = ""
    depth_stack = []
    n_str = len(latex_str)
    idx = 0
    while idx < n_str:
        found_left_delim = False
        for l_delim_str, r_delim_str in ADDITIVE_TERMS_LEFT_DELIMS.items():
            if latex_str.startswith(l_delim_str, idx):
                depth_stack.append(r_delim_str)
                buffer += latex_str[idx : idx + len(l_delim_str)]
                idx += len(l_delim_str)
                found_left_delim = True
                break
        if found_left_delim: continue

        found_right_delim = False
        if depth_stack:
            expected_closer = depth_stack[-1]
            if expected_closer.startswith("\\right") and latex_str.startswith(expected_closer, idx):
                depth_stack.pop()
                buffer += latex_str[idx : idx + len(expected_closer)]
                idx += len(expected_closer)
                found_right_delim = True
        if found_right_delim: continue

        char = latex_str[idx]

        if char in OPENING_BRACKETS_FOR_DEPTH:
            if char == '(': depth_stack.append(')')
            elif char == '[': depth_stack.append(']')
            elif char == '{': depth_stack.append('}')
            buffer += char
            idx += 1
        elif char in CLOSING_BRACKETS_FOR_DEPTH:
            if depth_stack and depth_stack[-1] == char: depth_stack.pop()
            buffer += char
            idx += 1
        elif latex_str.startswith("\\frac", idx) and not depth_stack:
            buffer += char
            idx += 1
        elif char in ('+', '-') and not depth_stack:
            if buffer: terms.append(buffer.strip())
            buffer = char
            idx += 1
        else:
            buffer += char
            idx += 1

    if buffer: terms.append(buffer.strip())

    final_terms = []
    if not terms: return None

    if terms[0] == "" and len(terms) > 1:
        first_term_candidate = terms[1]
        if first_term_candidate.startswith('+') or first_term_candidate.startswith('-'):
            final_terms.append(first_term_candidate)
            start_idx_for_loop = 2
        else:
            final_terms.append(terms[0])
            final_terms.append(terms[1])
            start_idx_for_loop = 2
    elif terms[0] != "" and (terms[0].startswith('+') or terms[0].startswith('-')):
        final_terms.append(terms[0])
        start_idx_for_loop = 1
    else:
        final_terms.append(terms[0])
        start_idx_for_loop = 1

    for i in range(start_idx_for_loop, len(terms)):
        final_terms.append(terms[i])

    final_terms = [t for t in final_terms if t or t == terms[0]]
    if final_terms and final_terms[0] == "" and len(final_terms) > 1:
        final_terms.pop(0)

    return final_terms if len(final_terms) >= 2 else None


# --- Part 2: Wrap with aligned if needed (MODIFIED) ---
DEFAULT_MAX_TERMS_FOR_ALIGN = 2
DEFAULT_MAX_LEN_FOR_ALIGN = 160
DEFAULT_MAX_LINE_LEN_IN_ALIGNED = 100 # New default constant

def wrap_with_aligned_if_needed(latex_str,
                                 max_terms_trigger=DEFAULT_MAX_TERMS_FOR_ALIGN,
                                 max_len_trigger=DEFAULT_MAX_LEN_FOR_ALIGN,
                                 max_line_len=DEFAULT_MAX_LINE_LEN_IN_ALIGNED): # New parameter
    terms = get_additive_terms(latex_str)

    apply_aligned_wrapper = False
    if terms:
        if len(terms) > max_terms_trigger or len(latex_str) > max_len_trigger:
            apply_aligned_wrapper = True

    if not apply_aligned_wrapper:
        return latex_str

    if not terms: # Should be caught by apply_aligned_wrapper logic, but safeguard
        return latex_str

    output_lines_as_term_lists = [] # List of lists of terms
    current_line_term_group = []

    # Add first term to the first group
    current_line_term_group.append(terms[0].strip())

    for i in range(1, len(terms)):
        term_with_sign = terms[i].strip()

        # Construct the string for the current line *if this term were added*
        cur_line_length = sum(len(term) for term in current_line_term_group) if current_line_term_group else 0
        new_line_length = cur_line_length + len(term_with_sign) + (1 if current_line_term_group else 0)

        ratio_of_current = (cur_line_length + 1) / (max_line_len + 1) + (max_line_len + 1) / (cur_line_length + 1)
        ratio_of_extended = (new_line_length + 1) / (max_line_len + 1) + (max_line_len + 1) / (new_line_length + 1)
        # ratio_of_new_line = (len(term_with_sign) + 1) / (max_line_len + 1) + (max_line_len + 1) / (len(term_with_sign) + 1)

        if (cur_line_length > max_line_len or ratio_of_extended > ratio_of_current) and len(current_line_term_group) > 0:
            # The new term makes the line too long. Finalize the previous line.
            output_lines_as_term_lists.append(list(current_line_term_group))
            current_line_term_group = [term_with_sign] # Start new line with the current term
        else:
            # Term fits, add it to the current line group
            current_line_term_group.append(term_with_sign)

    # Add the last accumulated line group
    if current_line_term_group:
        output_lines_as_term_lists.append(list(current_line_term_group))

    # Format with \begin{aligned} and &
    aligned_body_parts = []
    for i, term_group in enumerate(output_lines_as_term_lists):
        # Construct line string: "term1" + " " + "+term2" ...
        line_str = term_group[0]
        for k in range(1, len(term_group)):
            line_str += " " + term_group[k]
        # if i == 0 and line_str.count("=") == 1:
        #     # TODO: Handle cases where there are multiple "="
        #     line_str = line_str.replace("=", "= & ")
        # else:
        #     # Each line starts with "& "
        line_str = "& " + line_str
        aligned_body_parts.append(line_str)

    if len(aligned_body_parts) == 1:
        return latex_str

    aligned_body = "\\\\\n".join(aligned_body_parts)

    return f"\\begin{{aligned}}{aligned_body}\\end{{aligned}}"


# --- Part 3: Recursive Formatting (MODIFIED to pass new param) ---
def _find_next_opener_info(latex_str, start_pos, opener_regex_patterns):
    first_match_pos = -1
    found_actual_opener, found_opener_regex = None, None
    for opener_pattern in opener_regex_patterns:
        try:
            match_iter = re.finditer(opener_pattern, latex_str[start_pos:])
            for match in match_iter:
                current_match_pos = start_pos + match.start()
                if first_match_pos == -1 or current_match_pos < first_match_pos:
                    first_match_pos, found_actual_opener, found_opener_regex = \
                        current_match_pos, match.group(0), opener_pattern
                break
        except re.error: pass
    return (first_match_pos, found_actual_opener, found_opener_regex) if first_match_pos != -1 else (-1, None, None)

def recursive_latex_auto_linebreak(latex_str,
                           max_terms_aligned=DEFAULT_MAX_TERMS_FOR_ALIGN,
                           max_len_aligned=DEFAULT_MAX_LEN_FOR_ALIGN,
                           max_line_len_in_aligned=DEFAULT_MAX_LINE_LEN_IN_ALIGNED): # New parameter
    processed_parts = []
    current_pos = 0
    n = len(latex_str)

    while current_pos < n:
        match_pos, actual_opener_text, opener_regex = _find_next_opener_info(latex_str, current_pos, OPENER_REGEX_PATTERNS)

        if match_pos == -1:
            processed_parts.append(latex_str[current_pos:])
            break

        if match_pos > current_pos:
            processed_parts.append(latex_str[current_pos:match_pos])

        processed_parts.append(actual_opener_text)
        current_pos = match_pos + len(actual_opener_text)

        if opener_regex == r"\\frac":
            num_content, end_pos_num = _get_latex_argument_content(latex_str, current_pos)
            # Pass all formatting parameters recursively
            formatted_num = recursive_latex_auto_linebreak(num_content, max_terms_aligned, max_len_aligned, max_line_len_in_aligned)
            processed_parts.append("{" + formatted_num + "}")
            current_pos = end_pos_num

            den_content, end_pos_den = _get_latex_argument_content(latex_str, current_pos)
            formatted_den = recursive_latex_auto_linebreak(den_content, max_terms_aligned, max_len_aligned, max_line_len_in_aligned)
            processed_parts.append("{" + formatted_den + "}")
            current_pos = end_pos_den
        else:
            literal_closer = DELIM_PAIRS.get(actual_opener_text)
            if not literal_closer:
                continue

            content_start_idx = current_pos
            level, search_idx, end_delim_start_idx = 0, content_start_idx, -1

            while search_idx < n:
                if latex_str.startswith(actual_opener_text, search_idx):
                    level += 1
                    search_idx += len(actual_opener_text)
                elif latex_str.startswith(literal_closer, search_idx):
                    if level == 0:
                        end_delim_start_idx = search_idx
                        break
                    level -= 1
                    search_idx += len(literal_closer)
                elif latex_str.startswith("\\" + actual_opener_text, search_idx):
                    search_idx += len("\\" + actual_opener_text)
                elif latex_str.startswith("\\" + literal_closer, search_idx):
                    search_idx += len("\\" + literal_closer)
                else:
                    search_idx += 1

            if end_delim_start_idx != -1:
                content = latex_str[content_start_idx : end_delim_start_idx]
                # Pass all formatting parameters recursively
                formatted_content = recursive_latex_auto_linebreak(content, max_terms_aligned, max_len_aligned, max_line_len_in_aligned)
                processed_parts.append(formatted_content)
                processed_parts.append(literal_closer)
                current_pos = end_delim_start_idx + len(literal_closer)
            else:
                pass

    final_reconstructed_str = "".join(processed_parts)
    # Pass all formatting parameters to the final wrap
    return wrap_with_aligned_if_needed(final_reconstructed_str, max_terms_aligned, max_len_aligned, max_line_len_in_aligned)


# --- Example Usage ---
if __name__ == '__main__':
    print("--- Part 1: get_additive_terms (sanity check) ---")
    s1 = "4a+2b-3c/5+(c-a)^2/2"
    print(f"'{s1}' -> {get_additive_terms(s1)}")
    s_long_sum = "a+b+c+d+e+f+g+h+i"
    print(f"'{s_long_sum}' -> {get_additive_terms(s_long_sum)}")


    print("\n--- Part 2: wrap_with_aligned_if_needed (with new line length logic) ---")
    expr_to_wrap = "termA+termB+termC+termD+termE+termF+termG"
    print(f"\nInput: '{expr_to_wrap}'")
    print(f"Wrapped (max_terms=2, max_len=20, max_line_len=15):") # max_terms=2 means 3+ terms trigger align
    print(wrap_with_aligned_if_needed(expr_to_wrap, max_terms_trigger=2, max_len_trigger=20, max_line_len=15))
    # termA +termB (len approx 6+1+6=13) -> fits
    # termA +termB +termC (len approx 13+1+6=20) -> exceeds 15. Break.
    # Expected:
    # & termA +termB \\
    # & +termC +termD \\
    # & +termE +termF \\
    # & +termG

    print(f"\nInput: '{s_long_sum}' (a+b+c+d+e+f+g+h+i)")
    print(f"Wrapped (max_terms=2, max_len=5, max_line_len=10):") # triggers align, line length 10
    print(wrap_with_aligned_if_needed(s_long_sum, max_terms_trigger=2, max_len_trigger=5, max_line_len=10))
    # a (1)
    # a +b (1+1+2=4)
    # a +b +c (4+1+2=7)
    # a +b +c +d (7+1+2=10)
    # a +b +c +d +e (10+1+2=13) > 10. Break.
    # Line 1: a +b +c +d
    # New line starts with +e
    # +e (2)
    # +e +f (2+1+2=5)
    # +e +f +g (5+1+2=8)
    # +e +f +g +h (8+1+2=11) > 10. Break.
    # Line 2: +e +f +g
    # New line starts with +h
    # +h (2)
    # +h +i (2+1+2=5)
    # Line 3: +h +i
    # Expected:
    # & a +b +c +d\\
    # & +e +f +g\\
    # & +h +i

    print("\n--- Part 3: recursive_latex_auto_linebreak (passing new param) ---")
    # Test with a long expression inside parentheses, requiring internal alignment with line length
    long_paren_expr = "f(x) = (a+b+c+d+e+f+g+h+i+j+k+l+m+n+o+p+q+r) + (s+t+u+v+w) - 2(x+y+z)"
    print(f"\nInput: '{long_paren_expr}'")
    print(f"Recursively formatted (max_terms_aligned=2, max_len_aligned=30, max_line_len_in_aligned=20):")
    # First parenthesis content: "a+b+...+r" (18 terms) -> will be wrapped by aligned
    #   Inside that aligned: lines should be <= 20 chars.
    #   a +b +c +d +e +f +g +h +i (9 terms, length ~ 9*2 + 8*1 = 26) -> should break
    #   Example: a +b +c +d +e +f +g +h (8 terms, 8*2+7 = 23)
    #            a +b +c +d +e +f +g (7 terms, 7*2+6 = 20) -> fits 20
    # Outer expression: 3 terms: "(...)" + "(...)" - "2(...)" -> may be wrapped by aligned if total > 30 or terms > 2
    print(recursive_latex_auto_linebreak(long_paren_expr, max_terms_aligned=2, max_len_aligned=30, max_line_len_in_aligned=20))

    print("-" * 20)
    nested_frac_long_num = "E = \\frac{X_1+X_2+X_3+X_4+X_5+X_6+X_7+X_8+X_9+X_{10}}{k+l+m} + Y"
    print(f"\nInput: '{nested_frac_long_num}'")
    print(f"Recursively formatted (max_terms_aligned=1, max_len_aligned=10, max_line_len_in_aligned=18):")
    # Numerator has 10 terms. max_terms_aligned=1 means >=2 terms will trigger aligned.
    # max_line_len_in_aligned=18 for numerator's lines.
    # X_1 +X_2 +X_3 (len ~3+1+3+1+3 = 11)
    # X_1 +X_2 +X_3 +X_4 (len ~11+1+3 = 15)
    # X_1 +X_2 +X_3 +X_4 +X_5 (len ~15+1+3 = 19) > 18. Break.
    # Line 1: X_1 +X_2 +X_3 +X_4
    print(recursive_latex_auto_linebreak(nested_frac_long_num, max_terms_aligned=1, max_len_aligned=10, max_line_len_in_aligned=18))

File: static/js/linebreak.js
// latex_auto_linebreak.js
// Converted from the original Python file to run in browser (ES6).
// Provides functions:
//  - getAdditiveTerms(latexStr)
//  - wrapWithAlignedIfNeeded(latexStr, maxTermsTrigger, maxLenTrigger, maxLineLen)
//  - recursiveLatexAutoLinebreak(latexStr, maxTermsAligned, maxLenAligned, maxLineLenInAligned)
// Example usage at the bottom prints results to console.

// -------------------- Constants / Helpers --------------------
const DELIM_PAIRS = {
    "\\left(": "\\right)",
    "\\left[": "\\right]",
    "\\left{": "\\right}",
    "\\left|": "\\right|",
    "\\left.": "\\right.",
    "(": ")",
    "[": "]",
};

const OPENER_REGEX_PATTERNS = [
    /\\left\(/, /\\left\[/, /\\left\{/, /\\left\|/, /\\left\./,
    /\\frac/,
    /\(/, /\[/
];

const OPENING_BRACKETS_FOR_DEPTH = ['(', '[', '{'];
const CLOSING_BRACKETS_FOR_DEPTH = [')', ']', '}'];

// Keep the additive-terms left delimiters in a deterministic array order
const ADDITIVE_TERMS_LEFT_DELIMS_ARR = [
    ["\\left(", "\\right)"],
    ["\\left[", "\\right]"],
    ["\\left{", "\\right}"],
    ["\\left|", "\\right|"],
    ["\\left.", "\\right."]
];

function _startsWithAt(s, sub, pos) {
    return s.substr(pos, sub.length) === sub;
}

// Attempt to extract one LaTeX 'argument' starting at startPos.
// Mirrors Python semantics:
//  - If it starts with '{', return the content inside matching braces (not including outer braces) and position after closing brace.
//  - If it starts with '\something' return that token (like \alpha or \sin or single-char escape) and new position.
//  - Otherwise return the single character and new position (startPos+1).
function _getLatexArgumentContent(text, startPos) {
    const n = text.length;
    if (startPos >= n) return { content: "", newPos: startPos };

    const ch = text[startPos];

    if (ch === "{") {
        let braceDepth = 0;
        for (let i = startPos; i < n; ++i) {
            if (text[i] === "{" && (i === 0 || text[i - 1] !== "\\")) {
                braceDepth += 1;
            } else if (text[i] === "}" && (i === 0 || text[i - 1] !== "\\")) {
                braceDepth -= 1;
                if (braceDepth === 0) {
                    return { content: text.slice(startPos + 1, i), newPos: i + 1 };
                }
            }
        }
        throw new Error(`Unmatched brace starting at ${startPos} in '${text}'`);
    } else if (ch === "\\") {
        const tail = text.slice(startPos);
        // Try to match a multi-letter command optionally followed by space and * or !
        const multiCmd = tail.match(/^\\[a-zA-Z]+(?:\s*[*!]?)?/);
        if (multiCmd) {
            const token = multiCmd[0];
            return { content: token, newPos: startPos + token.length };
        }
        // Fallback: single escaped char like \{
        const single = tail.match(/^\\./);
        if (single) {
            const token = single[0];
            return { content: token, newPos: startPos + token.length };
        }
        // Nothing matched, return the slash itself
        return { content: text[startPos], newPos: startPos + 1 };
    }
    return { content: text[startPos], newPos: startPos + 1 };
}

// -------------------- Part 1: Identify Additive Expressions --------------------
function getAdditiveTerms(latexStr) {
    if (typeof latexStr !== "string") return null;
    const terms = [];
    let buffer = "";
    const depthStack = [];
    const nStr = latexStr.length;
    let idx = 0;

    while (idx < nStr) {
        // check left delimiters (the special \left... forms)
        let foundLeftDelim = false;
        for (const [lDelimStr, rDelimStr] of ADDITIVE_TERMS_LEFT_DELIMS_ARR) {
            if (_startsWithAt(latexStr, lDelimStr, idx)) {
                depthStack.push(rDelimStr);
                buffer += latexStr.substr(idx, lDelimStr.length);
                idx += lDelimStr.length;
                foundLeftDelim = true;
                break;
            }
        }
        if (foundLeftDelim) continue;

        // check right delim for the last left \left... if any
        let foundRightDelim = false;
        if (depthStack.length > 0) {
            const expectedCloser = depthStack[depthStack.length - 1];
            if (expectedCloser.startsWith("\\right") && _startsWithAt(latexStr, expectedCloser, idx)) {
                depthStack.pop();
                buffer += latexStr.substr(idx, expectedCloser.length);
                idx += expectedCloser.length;
                foundRightDelim = true;
            }
        }
        if (foundRightDelim) continue;

        const char = latexStr[idx];

        if (OPENING_BRACKETS_FOR_DEPTH.includes(char)) {
            if (char === '(') depthStack.push(')');
            else if (char === '[') depthStack.push(']');
            else if (char === '{') depthStack.push('}');
            buffer += char;
            idx += 1;
        } else if (CLOSING_BRACKETS_FOR_DEPTH.includes(char)) {
            if (depthStack.length && depthStack[depthStack.length - 1] === char) depthStack.pop();
            buffer += char;
            idx += 1;
        } else if (_startsWithAt(latexStr, "\\frac", idx) && depthStack.length === 0) {
            // Note: original Python appended only the single char at idx in this case.
            buffer += char;
            idx += 1;
        } else if ((char === '+' || char === '-') && depthStack.length === 0) {
            if (buffer) terms.push(buffer.trim());
            buffer = char;
            idx += 1;
        } else {
            buffer += char;
            idx += 1;
        }
    }

    if (buffer) terms.push(buffer.trim());

    let finalTerms = [];
    if (!terms || terms.length === 0) return null;

    if (terms[0] === "" && terms.length > 1) {
        const firstTermCandidate = terms[1];
        if (firstTermCandidate && (firstTermCandidate.startsWith('+') || firstTermCandidate.startsWith('-'))) {
            finalTerms.push(firstTermCandidate);
            var startIdxForLoop = 2;
        } else {
            finalTerms.push(terms[0]);
            finalTerms.push(terms[1]);
            var startIdxForLoop = 2;
        }
    } else if (terms[0] !== "" && (terms[0].startsWith('+') || terms[0].startsWith('-'))) {
        finalTerms.push(terms[0]);
        var startIdxForLoop = 1;
    } else {
        finalTerms.push(terms[0]);
        var startIdxForLoop = 1;
    }

    for (let i = startIdxForLoop; i < terms.length; ++i) {
        finalTerms.push(terms[i]);
    }

    // Filter out empty terms except maybe keep the first original empty as Python logic did
    finalTerms = finalTerms.filter((t, idx) => (t || idx === 0));

    if (finalTerms.length && finalTerms[0] === "" && finalTerms.length > 1) {
        finalTerms.shift();
    }

    return finalTerms.length >= 2 ? finalTerms : null;
}

// -------------------- Part 2: Wrap with aligned if needed --------------------
const DEFAULT_MAX_TERMS_FOR_ALIGN = 2;
const DEFAULT_MAX_LEN_FOR_ALIGN = 160;
const DEFAULT_MAX_LINE_LEN_IN_ALIGNED = 100;

function wrapWithAlignedIfNeeded(
    latexStr,
    maxTermsTrigger = DEFAULT_MAX_TERMS_FOR_ALIGN,
    maxLenTrigger = DEFAULT_MAX_LEN_FOR_ALIGN,
    maxLineLen = DEFAULT_MAX_LINE_LEN_IN_ALIGNED
) {
    const terms = getAdditiveTerms(latexStr);

    let applyAlignedWrapper = false;
    if (terms) {
        if (terms.length > maxTermsTrigger || latexStr.length > maxLenTrigger) {
            applyAlignedWrapper = true;
        }
    }

    if (!applyAlignedWrapper) return latexStr;
    if (!terms) return latexStr; // safeguard

    const outputLinesAsTermLists = [];
    let currentLineTermGroup = [];

    // add first term to first group
    currentLineTermGroup.push(terms[0].trim());

    for (let i = 1; i < terms.length; ++i) {
        const termWithSign = terms[i].trim();

        // current line length as sum of lengths of terms
        const curLineLength = currentLineTermGroup.reduce((acc, t) => acc + t.length, 0);
        const newLineLength = curLineLength + termWithSign.length + (currentLineTermGroup.length ? 1 : 0);

        // compute the ratio heuristics used in Python
        const ratioOfCurrent = (curLineLength + 1) / (maxLineLen + 1) + (maxLineLen + 1) / (curLineLength + 1);
        const ratioOfExtended = (newLineLength + 1) / (maxLineLen + 1) + (maxLineLen + 1) / (newLineLength + 1);

        if ((curLineLength > maxLineLen || ratioOfExtended > ratioOfCurrent) && currentLineTermGroup.length > 0) {
            outputLinesAsTermLists.push(Array.from(currentLineTermGroup));
            currentLineTermGroup = [termWithSign];
        } else {
            currentLineTermGroup.push(termWithSign);
        }
    }

    if (currentLineTermGroup.length) {
        outputLinesAsTermLists.push(Array.from(currentLineTermGroup));
    }

    // Build aligned body parts
    const alignedBodyParts = [];
    for (let i = 0; i < outputLinesAsTermLists.length; ++i) {
        const termGroup = outputLinesAsTermLists[i];
        let lineStr = termGroup[0];
        for (let k = 1; k < termGroup.length; ++k) {
            lineStr += " " + termGroup[k];
        }
        lineStr = "& " + lineStr;
        alignedBodyParts.push(lineStr);
    }

    if (alignedBodyParts.length === 1) return latexStr;

    const alignedBody = alignedBodyParts.join("\\\\\n");
    return `\\begin{aligned}${alignedBody}\\end{aligned}`;
}

// -------------------- Part 3: Recursive Formatting --------------------
function _findNextOpenerInfo(latexStr, startPos, openerRegexPatterns) {
    const substring = latexStr.slice(startPos);
    let firstMatchPos = -1;
    let foundActualOpener = null;
    let foundOpenerRegex = null;

    for (const pattern of openerRegexPatterns) {
        // Use pattern without global state (the RegExp object may be reused)
        const re = new RegExp(pattern.source);
        const m = re.exec(substring);
        if (m) {
            const currentMatchPos = startPos + m.index;
            if (firstMatchPos === -1 || currentMatchPos < firstMatchPos) {
                firstMatchPos = currentMatchPos;
                foundActualOpener = m[0];
                foundOpenerRegex = pattern; // keep the original pattern object for reference
            }
        }
    }

    if (firstMatchPos === -1) return { matchPos: -1, actualOpenerText: null, openerRegex: null };
    return { matchPos: firstMatchPos, actualOpenerText: foundActualOpener, openerRegex: foundOpenerRegex };
}

function recursiveLatexAutoLinebreak(
    latexStr,
    maxTermsAligned = DEFAULT_MAX_TERMS_FOR_ALIGN,
    maxLenAligned = DEFAULT_MAX_LEN_FOR_ALIGN,
    maxLineLenInAligned = DEFAULT_MAX_LINE_LEN_IN_ALIGNED
) {
    const processedParts = [];
    let currentPos = 0;
    const n = latexStr.length;

    while (currentPos < n) {
        const { matchPos, actualOpenerText, openerRegex } = _findNextOpenerInfo(latexStr, currentPos, OPENER_REGEX_PATTERNS);

        if (matchPos === -1) {
            processedParts.push(latexStr.slice(currentPos));
            break;
        }

        if (matchPos > currentPos) {
            processedParts.push(latexStr.slice(currentPos, matchPos));
        }

        processedParts.push(actualOpenerText);
        currentPos = matchPos + actualOpenerText.length;

        if (actualOpenerText === "\\frac") {
            // numerator
            const { content: numContent, newPos: endPosNum } = _getLatexArgumentContent(latexStr, currentPos);
            const formattedNum = recursiveLatexAutoLinebreak(numContent, maxTermsAligned, maxLenAligned, maxLineLenInAligned);
            processedParts.push("{" + formattedNum + "}");
            currentPos = endPosNum;

            // denominator
            const { content: denContent, newPos: endPosDen } = _getLatexArgumentContent(latexStr, currentPos);
            const formattedDen = recursiveLatexAutoLinebreak(denContent, maxTermsAligned, maxLenAligned, maxLineLenInAligned);
            processedParts.push("{" + formattedDen + "}");
            currentPos = endPosDen;
        } else {
            const literalCloser = DELIM_PAIRS[actualOpenerText];
            if (!literalCloser) {
                // If no matching literal closer defined, continue scanning (replicates Python's continue)
                continue;
            }

            const contentStartIdx = currentPos;
            let level = 0;
            let searchIdx = contentStartIdx;
            let endDelimStartIdx = -1;

            while (searchIdx < n) {
                if (_startsWithAt(latexStr, actualOpenerText, searchIdx)) {
                    level += 1;
                    searchIdx += actualOpenerText.length;
                } else if (_startsWithAt(latexStr, literalCloser, searchIdx)) {
                    if (level === 0) {
                        endDelimStartIdx = searchIdx;
                        break;
                    }
                    level -= 1;
                    searchIdx += literalCloser.length;
                } else if (_startsWithAt(latexStr, "\\" + actualOpenerText, searchIdx)) {
                    // skip escaped opener
                    searchIdx += ("\\" + actualOpenerText).length;
                } else if (_startsWithAt(latexStr, "\\" + literalCloser, searchIdx)) {
                    // skip escaped closer
                    searchIdx += ("\\" + literalCloser).length;
                } else {
                    searchIdx += 1;
                }
            }

            if (endDelimStartIdx !== -1) {
                const content = latexStr.slice(contentStartIdx, endDelimStartIdx);
                const formattedContent = recursiveLatexAutoLinebreak(content, maxTermsAligned, maxLenAligned, maxLineLenInAligned);
                processedParts.push(formattedContent);
                processedParts.push(literalCloser);
                currentPos = endDelimStartIdx + literalCloser.length;
            } else {
                // unmatched case: just continue (same as Python 'pass')
            }
        }
    }

    const finalReconstructedStr = processedParts.join("");
    return wrapWithAlignedIfNeeded(finalReconstructedStr, maxTermsAligned, maxLenAligned, maxLineLenInAligned);
}

// -------------------- Example usage (prints to console) --------------------
function _exampleLatexAutoLinebreak() {
    console.log("--- Part 1: getAdditiveTerms (sanity check) ---");
    const s1 = "4a+2b-3c/5+(c-a)^2/2";
    console.log(`'${s1}' ->`, getAdditiveTerms(s1));
    const sLongSum = "a+b+c+d+e+f+g+h+i";
    console.log(`'${sLongSum}' ->`, getAdditiveTerms(sLongSum));

    console.log("\n--- Part 2: wrapWithAlignedIfNeeded (with new line length logic) ---");
    const exprToWrap = "termA+termB+termC+termD+termE+termF+termG";
    console.log(`\nInput: '${exprToWrap}'`);
    console.log("Wrapped (max_terms=2, max_len=20, max_line_len=15):");
    console.log(wrapWithAlignedIfNeeded(exprToWrap, 2, 20, 15));

    console.log(`\nInput: '${sLongSum}' (a+b+c+d+e+f+g+h+i)`);
    console.log("Wrapped (max_terms=2, max_len=5, max_line_len=10):");
    console.log(wrapWithAlignedIfNeeded(sLongSum, 2, 5, 10));

    console.log("\n--- Part 3: recursiveLatexAutoLinebreak (passing new param) ---");
    const longParenExpr = "f(x) = (a+b+c+d+e+f+g+h+i+j+k+l+m+n+o+p+q+r) + (s+t+u+v+w) - 2(x+y+z)";
    console.log(`\nInput: '${longParenExpr}'`);
    console.log("Recursively formatted (max_terms_aligned=2, max_len_aligned=30, max_line_len_in_aligned=20):");
    console.log(recursiveLatexAutoLinebreak(longParenExpr, 2, 30, 20));

    console.log("-".repeat(20));
    const nestedFracLongNum = "E = \\frac{X_1+X_2+X_3+X_4+X_5+X_6+X_7+X_8+X_9+X_{10}}{k+l+m} + Y";
    console.log(`\nInput: '${nestedFracLongNum}'`);
    console.log("Recursively formatted (max_terms_aligned=1, max_len_aligned=10, max_line_len_in_aligned=18):");
    console.log(recursiveLatexAutoLinebreak(nestedFracLongNum, 1, 10, 18));
}


// -------------------- Exports --------------------
// If used in module environment (e.g., bundler), export functions:
if (typeof module !== "undefined" && typeof module.exports !== "undefined") {
    module.exports = {
        getAdditiveTerms,
        wrapWithAlignedIfNeeded,
        recursiveLatexAutoLinebreak
    };
}
